killswitch.py

```diff
@@ -0,0 +1,189 @@
+        logger.warning(f"Failed to get kill switches, data was invalid: {e}")
+        return None
+
+    except (requests.exceptions.BaseHTTPError, requests.exceptions.ConnectionError) as e:
+        logger.warning(f"unable to connect to {target!r}: {e}")
+        return None
+
+    return data
+
+
+def parse_kill_switches(data: KILL_SWITCH_JSON_DICT) -> List[KillSwitch]:
+    """
+    Parse kill switch dict to List of KillSwitches.
+
+    :param data: dict containing raw killswitch data
+    :return: a list of all provided killswitches
+    """
+    last_updated = data['last_updated']
+    ks_version = data['version']
+    logger.info(f'Kill switches last updated {last_updated}')
+
+    if ks_version != 1:
+        logger.warning(f'Unknown killswitch version {ks_version}. Bailing out')
+        return []
+
+    kill_switches = cast(KILL_SWITCH_JSON, data['kill_switches'])
+    out: List[KillSwitch] = []
+
+    for idx, ks_data in enumerate(kill_switches):
+        try:
+            ver = semantic_version.Version(ks_data['version'])
+
+        except ValueError as e:
+            logger.warning(f'could not parse killswitch idx {idx}: {e}')
+            continue
+
+        ks = KillSwitch(version=ver, kills=cast(Dict[str, str], ks_data['kills']))
+        out.append(ks)
+
+    return out
+
+
+def get_kill_switches(target=DEFAULT_KILLSWITCH_URL) -> Optional[KillSwitchSet]:
+    """
+    Get a kill switch set object.
+
+    :param target: the URL to fetch the killswitch JSON from, defaults to DEFAULT_KILLSWITCH_URL
+    :return: the KillSwitchSet for the URL, or None if there was an error
+    """
+    if (data := fetch_kill_switches(target)) is None:
+        logger.warning('could not get killswitches')
+        return None
+
+    return KillSwitchSet(parse_kill_switches(data))
+
+
+active: KillSwitchSet = KillSwitchSet([])
+
+
+def setup_main_list():
+    """
+    Set up the global set of kill switches for querying.
+
+    Plugins should NOT call this EVER.
+    """
+    if (data := fetch_kill_switches()) is None:
+        logger.warning("Unable to fetch kill switches. Setting global set to an empty set")
+        return
+
+    global active
+    active = KillSwitchSet(parse_kill_switches(data))
+
+
+def get_disabled(id: str, *, version: semantic_version.Version = _current_version) -> DisabledResult:
+    """
+    Query the global KillSwitchSet for whether or not a given ID is disabled.
+
+    See KillSwitchSet#is_disabled for more information
+    """
+    return active.get_disabled(id, version=version)
+
+
+def is_disabled(id: str, *, version=_current_version) -> bool:
+    """Query the global KillSwitchSet#is_disabled method."""
+    return active.is_disabled(id, version=version)
+
+
+def get_reason(id: str, *, version=_current_version) -> str:
+    """Query the global KillSwitchSet#get_reason method."""
+    return active.get_reason(id, version=version)
+
+
+if __name__ == "__main__":
+    setup_main_list()
+    print(f'{_current_version=}')
```

Comment:
This is either "prototyping" and needs removing, or should be converted to logging.

Reply:
Was prototyping, explicitly in the main guard for that reason, I'll remove it before I mark this ready to merge